### PR TITLE
C++ and related tests

### PR DESCRIPTION
**Pull request by jforissier**
Test cases related to https://github.com/OP-TEE/optee_os/pull/3693. And two cleanup/improvement patches.

### CODE REVIEW

File: host/xtest/regression_1000.c [hunk 1]
```diff
@@ -1934,3 +1934,30 @@ static void xtest_tee_test_1025(ADBG_Case_t *c)
...
+	TEEC_Session session = { 0 };
+	uint32_t ret_orig = 0;
+
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+			xtest_teec_open_session(&session, &os_test_ta_uuid,
+				NULL, &ret_orig)))
+		return;
+
+	Do_ADBG_BeginSubCase(c, "Before dlopen()");
+	(void)ADBG_EXPECT_TEEC_SUCCESS(c,
```
Comment (jenswi-linaro) [inline host/xtest/regression_1000.c:1949]: Cast isn't needed.

File: host/xtest/regression_1000.c [hunk 2]
```diff
@@ -1907,3 +1907,30 @@ static void xtest_tee_test_1024(ADBG_Case_t *c)
...
 #endif /* CFG_CORE_TPM_EVENT_LOG */
+
+static void xtest_tee_test_1025(ADBG_Case_t *c)
+{
+	TEEC_Session session = { 0 };
+	uint32_t ret_orig = 0;
+
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+			xtest_teec_open_session(&session, &os_test_ta_uuid,
+				NULL, &ret_orig)))
```
Comment (etienne-lms) [inline host/xtest/regression_1000.c:1918]: could indent to `(`

File: host/xtest/regression_1000.c [hunk 3]
```diff
@@ -1934,3 +1934,30 @@ static void xtest_tee_test_1025(ADBG_Case_t *c)
...
+	Do_ADBG_BeginSubCase(c, "After dlopen()");
+	(void)ADBG_EXPECT_TEEC_SUCCESS(c,
+		TEEC_InvokeCommand(&session, TA_OS_TEST_CMD_DL_PHDR_DL, NULL,
+				   &ret_orig));
+	Do_ADBG_EndSubCase(c, "After dlopen()");
+
+	TEEC_CloseSession(&session);
+}
+ADBG_CASE_DEFINE(regression, 1026, xtest_tee_test_1026,
+		"Test dl_iterate_phdr()");
```
Comment (etienne-lms) [inline host/xtest/regression_1000.c:1963]: bad indentation. possibly also in previous commit(s).

File: ta/os_test/cxx_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ * Copyright (c) 2019 Huawei Technologies Co., Ltd
+ */
+
+
+extern "C" {
+
+#include <dlfcn.h>
+#include <stdint.h>
+#include <tee_ta_api.h>
+#include <os_test_lib.h>
```
Comment (etienne-lms) [inline ta/os_test/cxx_tests.cpp:12]: move 2 lines above

File: ta/os_test/cxx_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+			RTLD_NOW | RTLD_GLOBAL | RTLD_NODELETE);
+	if (!handle)
+		return TEE_ERROR_GENERIC;
+
+	ctor_test_fn = (TEE_Result (*)(void))dlsym(handle,
+						   "os_test_shlib_dl_cxx_ctor");
+	if (!ctor_test_fn)
+		goto err;
+	res = ctor_test_fn();
+err:
```
Comment (etienne-lms) [inline ta/os_test/cxx_tests.cpp:57]: maybe:
```c
	if (ctor_test_fn)
		res = ctor_test_fn();
```
unless you expect more tests here?

File: ta/os_test_lib/os_test_lib_cxx.cpp [hunk 6]
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Linaro Limited
```
Comment (etienne-lms) [inline ta/os_test_lib/os_test_lib_cxx.cpp:3]: 2018-2020?

File: ta/os_test/cxx_tests.h [hunk 7]
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2020 Huawei Technologies Co., Ltd
+ */
+
```
Comment (etienne-lms) [inline ta/os_test/cxx_tests.h:5]: header guards

File: ta/os_test/cxx_tests.h [hunk 8]
```diff
@@ -0,0 +1,8 @@
+// SPDX-License-Identifier: BSD-2-Clause
```
Comment (etienne-lms) [inline ta/os_test/cxx_tests.h:1]: header `/* */`

File: host/xtest/regression_1000.c (hunk 1 above)
Comment (jforissier) [reply host/xtest/regression_1000.c:1949]: WIll fix

File: host/xtest/regression_1000.c (hunk 2 above)
Comment (jforissier) [reply host/xtest/regression_1000.c:1918]: Will fix

File: host/xtest/regression_1000.c (hunk 3 above)
Comment (jforissier) [reply host/xtest/regression_1000.c:1963]: Will fix

File: ta/os_test/cxx_tests.cpp (hunk 4 above)
Comment (jforissier) [reply ta/os_test/cxx_tests.cpp:12]: yep

File: ta/os_test/cxx_tests.cpp (hunk 5 above)
Comment (jforissier) [reply ta/os_test/cxx_tests.cpp:57]: No more tests for now so I'll write as you suggest

File: ta/os_test_lib/os_test_lib_cxx.cpp (hunk 6 above)
Comment (jforissier) [reply ta/os_test_lib/os_test_lib_cxx.cpp:3]: Agreed

File: ta/os_test/cxx_tests.h (hunk 7 above)
Comment (jforissier) [reply ta/os_test/cxx_tests.h:5]: Will add them

File: ta/os_test/cxx_tests.h (hunk 8 above)
Comment (jforissier) [reply ta/os_test/cxx_tests.h:1]: OK

**Comment by jforissier** [comment]
Update.

**Comment by etienne-lms** [comment]
@jforissier, could you remove the stale label.
This is active (slowly but still).

**Comment by jforissier** [comment]
> @jforissier, could you remove the stale label.
> This is active (slowly but still).

Indeed and I will rebase the optee_os PR soon.

**Comment by jforissier** [comment]
Ping for the [Update] patch

File: ta/os_test/cxx_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2020 Huawei Technologies Co., Ltd
+ */
+
```
Comment (etienne-lms) [inline ta/os_test/cxx_tests.cpp:5]: remove empty line

File: ta/os_test/cxx_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+	handle = dlopen("b3091a65-9751-4784-abf7-0298a7cc35ba",
+			RTLD_NOW | RTLD_GLOBAL | RTLD_NODELETE);
+	if (!handle)
+		return TEE_ERROR_GENERIC;
+
+	ctor_test_fn = (TEE_Result (*)(void))dlsym(handle,
+						   "os_test_shlib_dl_cxx_ctor");
+	if (ctor_test_fn)
+		res = ctor_test_fn();
+err:
```
Comment (etienne-lms) [inline ta/os_test/cxx_tests.cpp:56]: remove the label.

File: ta/os_test/os_test.c [hunk 11]
```diff
@@ -1298,3 +1299,101 @@ TEE_Result ta_entry_client_identity(uint32_t param_types, TEE_Param params[4])
...
 }
+
+__thread int os_test_tls_a;
+__thread int os_test_tls_b = 42;
+
+TEE_Result ta_entry_tls_test_main(void)
+{
+        if (os_test_tls_a != 0) {
+		EMSG("os_test_tls_a=%d, expected 0", os_test_tls_a);
+                return TEE_ERROR_GENERIC;
```
Comment (etienne-lms) [inline ta/os_test/os_test.c:1310]: fix indentation, here and below and the 3 other occurrences below.

File: ta/os_test/cxx_tests.cpp (hunk 9 above)
Comment (jforissier) [reply ta/os_test/cxx_tests.cpp:5]: ok

File: ta/os_test/cxx_tests.cpp (hunk 10 above)
Comment (jforissier) [reply ta/os_test/cxx_tests.cpp:56]: ok

File: ta/os_test/os_test.c (hunk 11 above)
Comment (jforissier) [reply ta/os_test/os_test.c:1310]: ok

**Comment by jforissier** [comment]
Thanks

**Pull request merged**